keps/sig-cli/3515-kubectl-explain-openapiv3/README.md

```diff
@@ -0,0 +1,909 @@
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+### OpenAPI v3 is a richer API description than OpenAPI v2
+
+OpenAPI v3 support in Kubernetes is currently beta since version 1.24.
+OpenAPI V3 is a richer representation of the kubernetes API to our users, who have been asking for visibility
+into things like:
+
+1. nullable
+2. default
+3. validation fields like oneOf, anyOf, etc.
+
+To show each of these additional data points by themselves is a strong reason
+to switch to using OpenAPI v3.
+
+
+### CRD schemas expressed as OpenAPI v2 are lossy
+
+Today CRDs specify their schemas in OpenAPI v3 format. To serve the `/openapi/v2`
+document used today by kubectl, there is an expensive conversion from the v3 down
+to v2 format.
+
+This process is [very lossy](https://github.com/kubernetes/kubernetes/blob/6e0de20fbb4c127d2e45c7a22347c08545fc7a86/staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/v2/conversion.go#L56-L66), so `kubectl explain` when used against CRDs
+making use of v3 features does not have a good experience with inaccurate information, or fields removed altogther.
+
+This transformation causes bugs, for example, when attempting to `explain` a field
+that is `nullable`, kubectl instead shows nothing, due to the lossy conversion
+wiping nullable fields.
+
+### Goals
+
+1. Provide the new richer type information specified by OpenAPI v3 within kubectl explain
+2. Have a more maintainable `text/template` based approach to printing
+3. Fallback to old `explain` implementation if cluster does not expose OpenAPI v3 data.
+4. Provide multiple new output formats for kubectl explain:
+    * human-readable plaintext
+    * markdown
+    * maybe others
+5. (Optional?) Allow users to specify their own templates for use with kubectl
+  explain (there may be interesting use cases for this)
+
+### Non-Goals
+
+1. "Fix" openapi v3 to openapi v2 conversion
+  This is a non-goal for two reasons:
+    * These formats are not compatible, and there WILL be data loss and inaccuracy
+    * This negates the benefits of using OpenAPI v3 for the richer type information
+2. Provide general-purpose OpenAPI visualization.
+
+
+## Proposal
+
+
+### Basic Usage
+
+The following user experience should be possible with `kubectl explain`
+
+```shell
+kubectl explain pods
+```
+
+Output should be familiar to users of today's `kubectl explain`, except new
+information is populated:
+
+```console
+<TODO>
+```
+
+Note: Feature during development will be gated by an experimental flag. The commands
+shown here elide the experimental flag for clarity.
+
+### Built-in Template Options
+#### Plaintext
+
+
+```shell
+kubectl explain pods
+```
+or
+
+```shell
+kubectl explain pods --output plaintext
+```
+
+The plaintext output format is the default and should be crafted to be as close
+as the existing `explain` output in use before this KEP.
+
+#### Raw
+
+```shell
+kubectl explain pods --output raw
```

Comment:
I would expect it outputs the raw openapi v4 data. The OpenAPI schema is versioned so I would expect the output to also contain the output version 4.0.0 for the `openapi` version field

Reply:
Yeah, I think we need to do one or the other of those things, so that clients have a way to tell how to parse the data and detect when/whether that needs to change. I'm leaning towards `--output openapiv3` because it gives us a path to explicitly deprecate the format, and even to support both formats concurrently during a transition period. The error message given to clients that fail to switch would also likely be more clear. The downside that comes to mind is that if the hypothetical v4 upgrade is completely transparent, we still need to go through the flag change process anyway.

(Fwiw I looked up existing uses of "raw" in kubectl, and it seems to be used to read/write exact bytes to kubeconfig files, and to override the URLs used for mutation commands.)

Reply:
To avoid overloading `raw` even further, I've changed the proposal to `--output openapiv3`